# Explore here

In [29]:
import pandas as pd 

# Carga de datos
Url = "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"
df = pd.read_csv(Url)

# Eliminar columnas irrelevantes 
df.drop("package_name", axis=1, inplace=True)

# Limpieza
df["review"] = df["review"].str.strip().str.lower()

print(df.head()) # Mira aquí los nombres de las columnas

                                              review  polarity
0  privacy at least put some option appear offlin...         0
1  messenger issues ever since the last update, i...         0
2  profile any time my wife or anybody has more t...         0
3  the new features suck for those of us who don'...         0
4  forced reload on uploading pic on replying com...         0


La previsualización de las primeras 5 filas valida que el dataset está limpio y listo para el proceso de Split.

In [35]:
from sklearn.model_selection import train_test_split

# Definimos 
X = df["review"]
y = df["polarity"] # Verifica si en tu df.head() sale 'polarity' o 'sentiment'


# Creación de conjuntos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print para verificar que las variables ya existen
print(f"Filas totales: {len(df)}")
print(f"X_train (80%): {len(X_train)}")
print(f"X_test (20%): {len(X_test)}")

Filas totales: 891
X_train (80%): 712
X_test (20%): 179


En este bloque se preparan los conjuntos para que el modelo pueda aprender y ser evaluado sin errores.

In [31]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

# Vectorización
vec_model = CountVectorizer(stop_words="english")
X_train_vec = vec_model.fit_transform(X_train).toarray()
X_test_vec = vec_model.transform(X_test).toarray()

# Modelos
gauss_model, bern_model, multi_model = GaussianNB(), BernoulliNB(), MultinomialNB()
gauss_model.fit(X_train_vec, y_train)
bern_model.fit(X_train_vec, y_train)
multi_model.fit(X_train_vec, y_train)

# Resultados
print(f"Precisión Gaussiana: {gauss_model.score(X_test_vec, y_test):.2%}")
print(f"Precisión Bernoulli: {bern_model.score(X_test_vec, y_test):.2%}")
print(f"Precisión Multinomial: {multi_model.score(X_test_vec, y_test):.2%}")

Precisión Gaussiana: 80.45%
Precisión Bernoulli: 77.09%
Precisión Multinomial: 81.56%


Se aplicó un split de 80% para entrenamiento y 20% para evaluación, asegurando una base de datos suficiente para el aprendizaje del modelo.

In [32]:
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

# Inicializar modelos
gauss_model = GaussianNB()
bern_model = BernoulliNB()
multi_model = MultinomialNB() 

# Entrenar y evaluar
gauss_model.fit(X_train_vec, y_train)
bern_model.fit(X_train_vec, y_train)
multi_model.fit(X_train_vec, y_train)

print(f"Precisión Gaussiana: {gauss_model.score(X_test_vec, y_test):.2%}")
print(f"Precisión Bernoulli: {bern_model.score(X_test_vec, y_test):.2%}")
print(f"Precisión Multinomial: {multi_model.score(X_test_vec, y_test):.2%}")

Precisión Gaussiana: 80.45%
Precisión Bernoulli: 77.09%
Precisión Multinomial: 81.56%


Tras comparar las tres implementaciones, el modelo MultinomialNB obtuvo el mejor rendimiento con un 81.56% de precisión.

In [33]:
from sklearn.ensemble import RandomForestClassifier

# Crear y entrenar Random Forest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train_vec, y_train)

# Evaluación
rf_acc = rf_model.score(X_test_vec, y_test)
print(f"Precisión Random Forest: {rf_acc:.2%}")

Precisión Random Forest: 79.89%


Evidencia de Rendimiento: Se comparó el modelo Naive Bayes (81.56%) contra Random Forest (79.89%).

conclusión 

Tras un análisis comparativo, se seleccionó la Regresión Logística como el modelo definitivo para producción, alcanzando una precisión de 83.24%.

In [34]:
from sklearn.linear_model import LogisticRegression
import joblib

# Entrenamiento del modelo ganador
log_model = LogisticRegression()
log_model.fit(X_train_vec, y_train)

# Evaluación de precisión
log_acc = log_model.score(X_test_vec, y_test)
print(f"Precisión Final: {log_acc:.2%}")

# Guardado de archivos 
joblib.dump(log_model, 'best_model.pkl')
joblib.dump(vec_model, 'vectorizer.pkl')

print("Archivos guardados correctamente.")

Precisión Final: 83.24%
Archivos guardados correctamente.


Este modelo superó al Naive Bayes (81.56%) y al Random Forest (79.89%), consolidándose como la mejor opción para este conjunto de datos.